Convert bufr into metatask

WalterKolczynski-NOAA opened the pull request:
# Description
Converts the atmos bufr sounding task to a metatask to process one forecast hour at a time. The hour to be processed is exported as a zero-padded string `FHR3`.
    
Also adds a (currently empty) collect job to create tarballs and do any other processing that needs to occur after all forecast hours have been processed. A stub `scripts/exgfs_atmos_bufr_sounding_collect.sh` script is created for the new job. All the old code that needs to be done once all forecast hours are complete will need to be moved there.
    
Refs: NOAA-EMC/global-workflow#1257


# Type of change
- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- Workflow creation only

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

BoCui-NOAA committed:
Add restart capability to job JGFS_ATMOS_POSTSND

BoCui-NOAA committed:
Update job JGFS_ATMOS_POSTSND package

BoCui-NOAA committed:
Update job JGFS_ATMOS_POSTSND package

BoCui-NOAA committed:
update exgfs_atmos_postsnd.sh to use NUM_SND_COLLECTIVES

WalterKolczynski-NOAA committed:
Convert bufr into metatask

Converts the atmos bufr sounding task to a metatask to process one
forecast hour at a time. The hour to be processed is exported as a
zero-padded string `FHR3`.

Also adds a (currently empty) collect job
to create tarballs and do any other processing that needs to occur
after all forecast hours have been processed.

Refs: NOAA-EMC/global-workflow#1257

WalterKolczynski-NOAA committed:
Remove extra blank line

WalterKolczynski-NOAA commented:
@BoCui-NOAA I'm on leave for two weeks starting Monday, so if you have any questions/comments on this PR or what's needed for your PR, please try to reach out soon.

BoCui-NOAA commented:
@WalterKolczynski-NOAA Thanks for the update. From the PR description, the goal of the job JGFS_ATMOS_BUFR_SOUNDING is to complete the processing of all forecast hours, one forecast hour at a time. The output files of this job will be temporary files saved in /ptmp. The job JGFS_ATMOS_BUFR_SOUNDING_COLLECT job is the final step to use the temporary files and complete the generation of the buffer sounding product.

Please correct me if I am wrong. Thanks!

WalterKolczynski-NOAA commented:
Correct, although there will need to be a couple modifications to keep the bufr temp files on stmp until they are collected instead of being put in a random directory and erased upon completion of the individual hour jobs.

WalterKolczynski-NOAA committed:
Add DATAoutput for sounding job

A known, temporary space is needed to hold intermediate bufr data
in temporary space until the collect job completes the process.
Following the model of the forecast job, a DATAoutput is defined
with a known path to store the data in until it has been processed.

WalterKolczynski-NOAA commented:
@BoCui-NOAA Just added a new bit so there is a space to store the intermediate bufr files until the are no longer needed. The individual forecast hour jobs should write them to `${DATAoutput}`, which is a separate temporary space that won't be scrubbed when the job ends. Then the collect job should read them from `${DATAoutput}`. Upon successful completion, the collect job will then remove `${DATAoutput}`

BoCui-NOAA commented:
@WalterKolczynski-NOAA Thank you for the prompt update.

The current bufr sounding package is parallel programming code that process multiple hours' files simultaneously not for a sing forecast hour. For example, it reads in the files from 000 to 012 hours and outputs a temporary product file containing the information from 000hr to 012hr. Then start another time period from 013hr to 024hr and output another temporary output file. The job fails if run one forecast hour at a time. I will update scripts/exgfs_atmos_bufr_sounding.sh. Is it acceptable for a metatask to process several forecast hours at a time? Thanks!

WalterKolczynski-NOAA commented:
If necessary, but it's something we've been trying to remove from workflow because it makes calculating dependencies more difficult. It would be preferable for the source code to be adapted (things like how many hours to process probably shouldn't be hard-coded anyway).

BoCui-NOAA commented:
@WalterKolczynski-NOAA 

This bufr sounding code is originally designed to handle multiple files. The Input and output file names are hardcoded into the code and controlled by various parameters such as the number of processes ( fixed value), forecast hours, and forecast interval hours etc. The MPI tasks assignment is set to vary for different forecast hours.

I made some adjustments for the code, letting it process just one file instead. The program runs, but gets stuck in infinite loops. It seems there are still several issues with processing only one file. We need to figure out how to assign tasks, balance the workload, and how processes communicate—all of these need to be carefully studied to solve the problems, and the final step, output evaluation, is also required.

Adjusting the code may take longer than expected. Now I am working on the bufr sounding update for the new GFS variables. I am afraid I do not have enough time to fix the problem to process one input file at the moment.

I suggest holding the code modification and choosing either one of the below options.

Option 1, add the restart capability. I have tested the package,  which is ready for the PR review and approval.

option 2,  keep the current codes. As you suggested, break the job into two. The first one is to process all the forecast hours and a metatask will process several forecast hours at a time. The second job is to create tarballs and finish all the other processing.

For this option, the job script needs some adjustments. Nested mpiexec commands are needed to achieve running multiple MPI jobs simultaneously for different forecast periods.

Any suggestions and comments? Thanks!

BoCui-NOAA commented:
@WalterKolczynski-NOAA Do you have experience using nested mpiexec commands? A sample would be helpful for me to structure the script.  Thanks!
